contentcuration/contentcuration/models.py

```diff
@@ -2225,3 +2216,74 @@ class Task(models.Model):
+    # For those changes related to users, store a user value instead of channel
+    # this may be different to created_by, as changes to invitations affect individual users.
+    # Indexed by default because it's a ForeignKey field.
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, on_delete=models.CASCADE, related_name="changes_about_user")
+    # Use client_rev to keep track of changes coming from the client side
+    # but let it be blank or null for changes we generate on the server side
+    client_rev = models.IntegerField(null=True, blank=True)
+    # client_rev numbers are by session, we add the session key here for bookkeeping
+    # to allow a check within the same session to return whether a change has been applied
+    # or not, and hence remove it from the frontend
+    session = models.ForeignKey(Session, null=True, blank=True, on_delete=models.SET_NULL)
+    table = models.CharField(max_length=32)
+    change_type = models.IntegerField()
+    # Use the DRF JSONEncoder class as the encoder here
+    # so that we can handle anything that has been deserialized by DRF
+    # or that will be later be serialized by DRF
+    kwargs = JSONField(encoder=JSONEncoder)
+    applied = models.BooleanField(default=False)
+    errored = models.BooleanField(default=False)
+
+    @classmethod
+    def _create_from_change(cls, created_by_id=None, channel_id=None, user_id=None, session_key=None, applied=False, table=None, rev=None, **data):
+        change_type = data.pop("type")
+        if table is None or table not in ALL_TABLES:
+            raise TypeError("table is a required argument for creating changes and must be a valid table name")
+        if change_type is None or change_type not in ALL_CHANGES:
+            raise TypeError("change_type is a required argument for creating changes and must be a valid change type integer")
+        return cls(
+            session_id=session_key,
+            created_by_id=created_by_id,
+            channel_id=channel_id,
+            user_id=user_id,
+            client_rev=rev,
+            table=table,
+            change_type=change_type,
+            kwargs=data,
+            applied=applied
+        )
+
+    @classmethod
+    def create_changes(cls, changes, created_by_id=None, session_key=None, applied=False):
+        change_models = []
+        for change in changes:
+            change_models.append(cls._create_from_change(created_by_id=created_by_id, session_key=session_key, applied=applied, **change))
+
+        cls.objects.bulk_create(change_models)
+        return change_models
+
+    @classmethod
+    def create_change(cls, change, created_by_id=None, session_key=None, applied=False):
+        obj = cls._create_from_change(created_by_id=created_by_id, session_key=session_key, applied=applied, **change)
+        obj.save()
+        return obj
+
+    def serialize_to_change_dict(self):
+        datum = self.kwargs
```

Comment:
Shoudl this be `self.kwargs.copy()`? Otherwise, what's the reasoning I'm missing for a separate variable?

Reply:
Yes, should be a copy - as otherwise we are modifying a property of the model in place!